services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/PathManagerService.java

```diff
@@ -18,18 +18,122 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Configuration;
 import org.springframework.stereotype.Service;
 
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
 @Service
+@Configuration
 public class PathManagerService {
+    @Autowired
+    private VpcSubnetsCache vpcSubnetsCache;
+
+    @Autowired
+    private SubnetPortsCacheV2 subnetPortsCache;
+
+    @Autowired
+    private VpcPathCache vpcPathCache;
+
     private static boolean USE_GRPC = true;
 
+    @Value("${path.mode}")
+    private String PATH_MODE;
+
+    @Value("${path.UPPER_VPC_SIZE}")
+    private int UPPER_VPC_SIZE;
+    @Value("${path.LOWER_VPC_SIZE}")
+    private int LOWER_VPC_SIZE;
+
     public boolean isFastPath() {
         return USE_GRPC;
     }
 
-    public boolean isFastPath(PortEntity portEntity) {
-        return USE_GRPC;
+    public boolean isFastPath(PortEntity portEntity) throws Exception{
+        switch (PATH_MODE) {
+            case "GRPC":
+                return USE_GRPC;
+            case "MQ":
+                return !USE_GRPC;
+            case "MIXED":
+                return choosePath(portEntity);
+            default:
+                throw new InvalidPathModeException();
+        }
+    }
+
+    private boolean choosePath(PortEntity portEntity) throws Exception{
+        if (portEntity == null) {
+            return USE_GRPC;
+        }
+
+        String vpcId = portEntity.getVpcId();
+
+        // Get VPC size - number of ports
+        AtomicInteger atomicNumberOfPorts = new AtomicInteger();
+        List<String> subnetIds;
+        subnetIds = vpcSubnetsCache.getVpcSubnets(vpcId).getSubnetIds();
+        subnetIds.stream().forEach(subnetId -> {
+            try {
+                InternalSubnetPorts internalSubnetPorts = subnetPortsCache.getSubnetPorts(subnetId);
+                atomicNumberOfPorts.addAndGet(internalSubnetPorts.getPorts().size());
+            } catch (Exception e) {
+
+            }
+        });
+
+        int numberOfPorts = atomicNumberOfPorts.get();
+
+        // Get the current path for this path
+        boolean currentPath;
+        try {
+            currentPath = vpcPathCache.getCurrentPathByVpcId(vpcId);
+        } catch (NullPointerException e) {
+            if (numberOfPorts > UPPER_VPC_SIZE) {
+                vpcPathCache.setPath(vpcId, USE_GRPC);
+                return USE_GRPC;
+            } else {
+                vpcPathCache.setPath(vpcId, !USE_GRPC);
+                return !USE_GRPC;
+            }
+
+        } catch (Exception e) {
+            throw e;
+        }
+
+        // Path switch logic
+        boolean chosenPath;
+
+        if (currentPath == USE_GRPC) {
+            if (numberOfPorts > UPPER_VPC_SIZE) {
+                chosenPath = !USE_GRPC;
```

Comment:
@VanderChen Why the line#105 shows that using `USE_GRPC` when `numberOfPorts > UPPER_VPC_SIZE`, but the line#120 shows that using `!USE_GRPC` in the same condition?

Reply:
@cj-chung Line#105 initialize a topic for a new VPC, and line#120 handles the situation of path change. As designed in [doc](https://github.com/futurewei-cloud/alcor/blob/master/docs/modules/ROOT/pages/mq_services/message_queue_system.adoc#channel-selection-algorithm), a simplified algorithm is implemented here.